### PR TITLE
tracing support

### PR DESCRIPTION
**Pull request by vyzo**
~Currently only the tracing scaffolding, but this should be enough for initial review pass.~

TBD:
- [x] implement tracer methods with particular event types
- [x] implement tracing to file
- [x] implement tracing to peer

### CODE REVIEW

**Review by Stebalien — DISMISSED** [review]
This seems too "not invented here".

**Comment by Stebalien** [comment]
(we discussed this on a call but I misunderstood some important context)

If we were trying to trace across machines, I'd understand maybe doing this ourselves. Given that we're _just_ tracing events locally, this seems way over the top. Why not just use Zap?

If the answer is "performance", have you tested that? I really doubt logging is going to be a bottleneck given encrypted transports, signing, verifying, etc.

**Comment by Stebalien** [comment]
Unless the concern is bandwidth. That's the one thing this buys us over a general-purpose logger like zap.

File: trace.go
Comment (Stebalien) [inline trace.go]: `pb.Message`?

**Comment by Stebalien** [comment]
Ok, I've discussed this with raul and looked at it a bit more. This still feels like a lot of code for something so simple but I understand why we're using protobufs (gRPC & bandwidth).

**Comment by vyzo** [comment]
Yes, we _are_ trying to trace across machines, it's just that only the local tracer has been written so far.
Tracing to peer (and gRPC based tracing) are forthcoming.
Also note, bandwidth is a real concern here.

File: trace.go
Comment (vyzo) [reply trace.go]: Yeah, we could do that to avoid the cast.

Comment (vyzo) [reply trace.go]: done, it takes a `*pb.TraceEvent` now.

**Comment by Stebalien** [comment]
> Yes, we are trying to trace across machines

I'm talking about including trace IDs in messages. However, message IDs should give us that anyways.

**Comment by vyzo** [comment]
> I'm talking about including trace IDs in messages. 

The peer ID of the originating peer is included in the message.

**Comment by vyzo** [comment]
Rebased on master.

File: floodsub.go
Comment (Stebalien) [inline floodsub.go:47]: Should this be in the _router_ or in the part that calls `AddPeer`?

File: floodsub.go
Comment (Stebalien) [inline floodsub.go:51]: ditto.

File: tracer.go
Comment (Stebalien) [inline tracer.go:1]: Double checking, this is moving to a new repo?

**Review by Stebalien — COMMENTED** [review]
After my initial "why are we doing this" reaction, this LGTM (modulo a few questions/nits).

File: floodsub.go
Comment (vyzo) [reply floodsub.go:47]: We could have it in the control substrate, it just felt more natural to do it in the router.
Other than that there is no particular reason for the choice.

File: tracer.go
Comment (vyzo) [reply tracer.go:1]: I think we can keep this file here, so that we have a one-stop construction of the pubsub system without requiring an external dependency.

File: floodsub.go
Comment (vyzo) [reply floodsub.go:51]: same, it could be in either place, just felt more natural to do it in the router.

File: tracer.go
Comment (Stebalien) [inline tracer.go:55]: Let's make this a bit safer.

File: tracer.go
Comment (Stebalien) [inline tracer.go]: We could end up with a pretty large buffer this way and/or drop messages. Is there a better way to do this?

File: tracer.go
Comment (Stebalien) [inline tracer.go]: Shouldn't we do this at this at the end?

File: tracer.go
Comment (Stebalien) [inline tracer.go]: Can't we just call open a stream? We shouldn't have to dial separately.

**Review by Stebalien — CHANGES_REQUESTED** [review]

File: tracer.go
Comment (vyzo) [reply tracer.go]: that's the buffer we are swapping in, it's already been consumed.

File: tracer.go
Comment (vyzo) [reply tracer.go]: We can, but the host will have to know how to reach the trace peer. The traced is not announcing itself to the dht, so it wouldn't be findable in general.

File: tracer.go
Comment (vyzo) [reply tracer.go:55]: Sure.

File: tracer.go
Comment (Stebalien) [reply tracer.go]: Yeah, but we could nil it out earlier to avoid holding on to the memory.

File: tracer.go
Comment (vyzo) [reply tracer.go]: We could check the size of the buffer and sleep only if the batch is smaller than a threshold.
We can also reduce the sleep time and poll a few times, say up to a second.

File: tracer.go
Comment (Stebalien) [reply tracer.go]: Ah. Shouldn't we add the the addresses to the peerstore on start with an expiration of "never" (https://github.com/libp2p/go-libp2p-core/blob/master/peerstore/peerstore.go#L45)?

File: tracer.go
Comment (vyzo) [reply tracer.go]: Oh, that's a good point. Ok, I'll move the niling out right after we write it.

Comment (vyzo) [reply tracer.go]: done.

File: tracer.go
Comment (vyzo) [reply tracer.go]: I added a check for the size of the batch, and polling every 100ms (for up to 1s) in order to accumulate buffer in safer way.

File: tracer.go
Comment (vyzo) [reply tracer.go]: that would work too; can do.

Comment (vyzo) [reply tracer.go]: done.

File: tracer.go
Comment (vyzo) [reply tracer.go:55]: I added a closed state variable, which gets set when we first Close, and is also consulted before adding to the event buffer.
It should be safe and idempotent now, no potential panics because of a closed channel.

File: tracer.go
Comment (Stebalien) [inline tracer.go]: `host.Peerstore().AddAddrs(pi.ID, pi.Addrs, ...)`?

Comment (Stebalien) [reply tracer.go]: (we can do them all at once)

File: tracer.go
Comment (Stebalien) [inline tracer.go:212]: for loop?

```go
t.mx.Lock()
for len(t.buf) < MinTraceBatchSize && time.Now().Before(deadline) { 
  t.mx.Unlock()
  time.Sleep(100 * time.Millisecond)
  t.mx.Lock()
}
```

**Review by Stebalien — APPROVED** [review]
nits but LGTM

File: tracer.go
Comment (vyzo) [reply tracer.go]: ah, good point; will do.

File: tracer.go
Comment (vyzo) [reply tracer.go:212]: the alergy to gotos strikes again! Ok, I'll write it as a for loop.

**Comment by raulk** [comment]
Hey, would’ve liked to review this properly before it was merged to master :-( Didn’t get a chance due to other things, but I’d have appreciated a mention and a heads-up warning that the merge train was leaving.

**Comment by vyzo** [comment]
@raulk you were extremely busy, so didn't want to burden you.
Feel free to make comments about potential improvements, and we can address them.
Note that the grpc trace client will live out of tree, as we don't want a dependency on grpc for mainline.

**Pull request merged**
